samza-sql/src/main/java/org/apache/samza/sql/translator/ScanTranslator.java

```diff
@@ -169,15 +174,27 @@ void translate(final TableScan tableScan, final String queryLogicalId, final Str
       return;
     }
 
-    KVSerde<Object, Object> noOpKVSerde = KVSerde.of(new NoOpSerde<>(), new NoOpSerde<>());
-    DelegatingSystemDescriptor
-        sd = systemDescriptors.computeIfAbsent(systemName, DelegatingSystemDescriptor::new);
-    GenericInputDescriptor<KV<Object, Object>> isd = sd.getInputDescriptor(streamId, noOpKVSerde);
+    // set the wrapper input transformer (SamzaSqlInputTransformer) in system descriptor
+    DelegatingSystemDescriptor sd = systemDescriptors.get(systemName);
+    if (sd == null) {
+      sd = new DelegatingSystemDescriptor(systemName,// null);
+          new SamzaSqlInputTransformer(null, true, true));
+      systemDescriptors.put(systemName, sd);
+    } else {
+      /* in SamzaSQL, there should be no systemDescriptor setup by user, so this branch happens only
+      * in case of Fan-OUT (i.e., same input stream used in multiple sql statements), or wehn same input
+      * used twice in same sql statement (e.g., select ... from input as i1, input as i2 ...), o.w., throw error */
+      if (sd.getTransformer().isPresent()) {
+        InputTransformer existingTransformer = sd.getTransformer().get();
+        if (!(existingTransformer instanceof SamzaSqlInputTransformer)) {
+          throw new SamzaException("SamzaSQL Exception: existing transformer for " + systemName + " is not SamzaSqlInputTransformer");
+        }
+      }
+    }
 
-    MessageStream<KV<Object, Object>> inputStream =
-        inputMsgStreams.computeIfAbsent(source, v -> streamAppDesc.getInputStream(isd));
+    InputDescriptor isd = sd.getInputDescriptor(streamId, new NoOpSerde<>(), sd.getTransformer().get());
```

Comment:
Is specifying transformer explicitly required? I think InputDescriptor should use the transformer for the SystemDescriptor it was obtained from by default.

Reply:
you are right, this is related to the above comment on getInputDescriptor taking InputTransformer as input change which I rolled back.